RFC: generic iterator based on stripes iterator

keegancsmith opened the pull request:
In a few places we have different patterns around pagination:
- Passing a closure in
- Returning a cursor for the client to use
- Just collect everything into a big slice

Other go libraries often will setup pagination patterns / use interfaces. The API I quite like is the stripe-go one which is minimal and reminds me a lot of Scanner from the stdlib. You just do something like

``` go
for it.Next() {
  doSomething(it.Current())
}
if it.Err() != nil {
  handle
}
```


This is a draft commit to show a single use of it in a PR I just wrote. However, I think this could apply to our DB interfaces nicely / we could adjust this pattern for different use cases.

This is just a draft so contains no tests/has bad naming (stutter)/etc. Just to start the conversation.

I imagine if we had this interface a few useful methods like

``` go
Map[A,B](func(A() B, Iterator[A]) Iterator[B]
Collect(Iterator[T]) ([]T, error)
```

etc. Basically could implement the same functions as present in https://pkg.go.dev/golang.org/x/exp/slices

Test Plan: n/a

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (+0.03 kb) | 0.03% (+3.95 kb) | 0.03% (+3.92 kb) | 0.14% (+1) 🔺 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/k/iterator/compare-report.html;tab=live_object) for a full comparison between the commits 132917c and c5181d0 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

bobheadxi commented on internal/iterator/iterator.go:
```diff
@@ -0,0 +1,42 @@
+package iterator
+
+func New[T any](next func() ([]T, error)) *Iterator[T] {
```
Should this go in `lib/` instead? That's where some of our other shared primitives like `lib/group` etc are, and we have a growing codebase over in Cloud that leverages `lib/` libraries outside of the monorepo 😁

keegancsmith replied:
Sounds good to me.

camdencheek commented on internal/iterator/iterator.go:
```diff
@@ -0,0 +1,42 @@
+package iterator
+
+func New[T any](next func() ([]T, error)) *Iterator[T] {
```
minor: why return a pointer here?

keegancsmith replied:
Because next is stateful, so we don't wanna copy the fields but keep the original references. IE you can get some weird footguns otherwise, but I suppose reading an iterator, then reading it from another function is asking for trouble :)

camdencheek replied:
Aah, that makes a lot of sense. Wasn't thinking of `next` as stateful.

camdencheek reviewed (approved):
I like this a lot. This is also my favorite iterator pattern. I particularly like that the generator function can generate a collection. That's a detail not included in other designs I've seen that makes it feel awkward to use in many common situations (like this example).

Strum355 commented on internal/iterator/iterator.go:
```diff
@@ -0,0 +1,42 @@
+	next func() ([]T, error)
+}
+
+func (it *Iterator[T]) Next() bool {
```
Im wondering is there a way to have this return (T, bool) while still working nicely with for loops 🤔 afk rn so I cant try stuff out

keegancsmith replied:
If you have a nice pattern here we could have another function which does it? But yeah, this pattern feels like it works nicely with what go gives ya.

Strum355 replied:
My quick 5 minute attempt, maybe theres something here you want to run with? Dont mind either way, Im a bit more of a fan of something built into for loops than a separate `Current()` method if you feel me https://go.dev/play/p/-0DrKoB7nP2

coury-clark replied:
I recently implemented a persistable iterator for code insights and used a similar pattern: https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/insights/scheduler/iterator/repo_iterator.go?L161

LawnGnome replied:
I've implemented a couple of ad hoc iterators for Batch Changes recently, and although they've used the return `(T, bool)` pattern, I think I'd basically come around to `Current`/`Next` as well, honestly. `T, bool` is great for getters, but not so good for iterators.

Strum355 replied:
Yea bit iffy with how long the for loop line becomes but i dont care either way

LawnGnome replied:
I think the idiomatic Go answer here would be `for v := it.C(); it.N(); v = it.C() { ... }`. :laughing:

Strum355 replied:
That would look good, just that C() only returns something after the first N() 🤔

LawnGnome replied:
Sorry, I was going for a lazy joke about method name length without writing actually correct code. :grimacing:

keegancsmith replied:
TBH this feels more clunky than the scanner like API. Hopefully one day we can hook into range then we can be happy.

I think you can make the API with `for v := it.C(); it.N(); v = it.C() { ... }.` work, but then suddenly the docstring for C becomes more complicated / opens up footguns.

My preference would be sticking the proposed simple scanner like API. Then you can add a helper which wraps it to make it possible to do nice forloops?

Strum355 replied:
Don't worry bout the joke lol, my approach just isnt great ultimately :laff:

keegancsmith commented:
Any suggestions on how to avoid stutter? I don't like that to name the iterator in code it will look like `*iterator.Iterator[T]`.

Anyone keen to try this out on some other code to see if this works? Or should I clean up and we can just mutate the API as we go along and learn.

vdavid commented on internal/iterator/iterator.go:
```diff
@@ -0,0 +1,42 @@
+}
+
+type Iterator[T any] struct {
+	current []T
```
I understand that the name `current` comes from the previous version of `iterator.go`, but I think it's a misnomer because:
 - it's a singular term for an array of things
 - we already use the term "current" for getting the first item of the array.

My suggestion is to call it `items` or `elements`.

vdavid reviewed (approved):
I love this; nice and clear! Suggested an internal name change.

mrnugget commented:
cc @thenamankumar who's been working on generalizing our graphql iterators.

LawnGnome reviewed (approved):
Big :+1: on this. I also strongly support including a few more generic helpers as well, as you suggested.

keegancsmith commented:
I've updated/responded to feedback, added an example and docstrings. I think this is now ready to review/land.

I think adding helpers can be follow up PRs and be motivated by concrete examples they will be used. So I have left those out.

vdavid commented:
> The only thing I don't like is the stutter. IE when you reference the iterator you call it `iterator.Iterator`. What do people think of maybe giving it a small name like they do in testing.T. EG we call it a `iterator.I` or `iterator.It`?

I think we can go with either. It's a decision that's very easy to change later.

eseliger commented on lib/iterator/iterator.go:
```diff
@@ -0,0 +1,57 @@
+	return !it.done
+}
+
+// Current returns the latest item advanced by Next. Note: this will panic if
+// Next returned false or if Next was never called.
+func (it *Iterator[T]) Current() T {
+	return it.items[0]
+}
```
can we check len first and panic with a nicer message so that it's easier to track what's happening?

eseliger commented:
> Edit: The only think I don't like is the stutter. IE when you reference the iterator you call it iterator.Iterator. What do people think of maybe giving it a small name like they do in testing.T. EG we call it a iterator.I or iterator.It?

No strong preference but a little stutter I never found terrible. I personally find it well readable, moreso than just `I` 🤷

eseliger commented on lib/iterator/iterator_test.go:
```diff
@@ -0,0 +1,29 @@
+	"github.com/sourcegraph/sourcegraph/lib/iterator"
+)
+
+func ExampleIterator() {
```
test case for err pls that shows that iteration is stopped before the end :)

eseliger commented on lib/iterator/iterator_test.go:
```diff
@@ -0,0 +1,29 @@
+	})
+
+	for it.Next() {
+		fmt.Printf("%d ", it.Current())
```
This should assert

keegancsmith replied:
its an example. What should it assert?

keegancsmith replied:
The output comment line does that for ya

eseliger replied:
Yes for the example that is true, but it would be great to have an actual test case, WDYT?

eseliger replied:
Or is this some Go magic I don't know about 🤔

keegancsmith replied:
examples run as part of test runs. So if the output of the example doesn't match you get a test failure. Nice go magic to get a free test and correct documentation.

eseliger replied:
oh damn!

vdavid replied:
Whoa, that was surprising and educational. For reference: https://go.dev/blog/examples

eseliger reviewed (approved):
LGTM, some comments on tests

keegancsmith commented:
@eseliger added nice panics and tests. Please take a look.

I went ahead and added two convenience function `From` and `Collect`. I decided to export these straight away because I imagine as soon as people start using this library, these functions will be helpful in tests.

eseliger reviewed (approved):
Nice!

unknwon commented:
👍
